tools/c7n_mailer/c7n_mailer/gcp/gcp_pubsub_processor.py

```diff
@@ -0,0 +1,97 @@
+# Copyright 2019 Capital One Services, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import six
+import json
+from c7n_mailer.email_delivery import EmailDelivery
+from c7n_gcp.client import Session
+import base64
+import zlib
+
+MAX_MESSAGES = 1000
+
+
+class MailerGcpPubSubProcessor(object):
+
+    def __init__(self, config, logger, session=None):
+        self.config = config
+        self.logger = logger
+        self.subscription = self.config['queue_url']
+        self.session = session or Session()
+
+    def run(self):
+        self.logger.info("Downloading messages from the GCP PubSub Subscription.")
+
+        # Get first set of messages to process
+        messages = self.receive_messages()
+
+        while len(messages) > 0:
+            # Discard_date is the timestamp of the last published message in the messages list
+            # and will be the date we need to seek to when we ack_messages
+            discard_date = messages['receivedMessages'][-1]['message']['publishTime']
```

Comment:
properly acking is preferred here imo this is kind of hacky, you can provide a list of ack ids so you can ack a bunch of messages at once

Reply:
Yeah this is actually kind of hacky. I struggled with this earlier in the year until I finally figured out how to handle this.  Pubsub is eventually consistent so what would happen is that if you ack'd a list of message id's...they would still be around the next time you retrieved messages. I'd have to keep an every growing list of acknowledged id's. Didn't seem too clean to me. Reading into Seek, it seems that is the recommended way of handling this use case using discard dates, etc. I'll have to find the article I read that convinced me of this. I'm going to upstream commit it like this and if push comes to shove I'll have to proof what I'm talking about to you guys.